Photo-Sharing-Sample/PhotoSharing/ViewModels/UserProfileViewModel.swift

```diff
@@ -45,105 +44,106 @@ extension UserProfileView {
             }
         }
 
         private func onReceive(event: DataStoreServiceEvent) {
             switch event {
             case .userSynced:
-                DispatchQueue.main.async {
-                    self.user = self.dataStoreService.user
-                }
+                user = dataStoreService.user
                 tryLoadPosts()
+
             case .postSynced:
                 dataStorePublisher?.cancel()
-                getNumberOfMyPosts()
-                fetchMyPosts(page: 0)
+
+                Task {
+                    await getNumberOfMyPosts()
+                    await fetchMyPosts(page: 0)
+                }
+
                 isPostSynced = true
             case .postCreated(let newPost):
-                DispatchQueue.main.async {
-                    self.loadedPosts.insert(newPost, at: 0)
+                loadedPosts.insert(newPost, at: 0)
+                Task {
+                    await getNumberOfMyPosts()
                 }
-                getNumberOfMyPosts()
+
             case .postDeleted(let post):
                 removePost(post)
-                getNumberOfMyPosts()
+                Task {
+                    await getNumberOfMyPosts()
+                }
+
             default:
                 break
             }
         }
 
-        func fetchMyPosts(page: Int) {
+        func fetchMyPosts(page: Int) async {
             guard user != nil else {
                 return
             }
 
             let predicateInput = Post.keys.postedBy == user?.id
             let sortInput = QuerySortInput.descending(Post.keys.createdAt)
             let paginationInput = QueryPaginationInput.page(UInt(page), limit: 10)
-            dataStoreService.query(Post.self,
-                                   where: predicateInput,
-                                   sort: sortInput,
-                                   paginate: paginationInput) {
-                switch $0 {
-                case .success(let posts):
-                    DispatchQueue.main.async {
-                        if page != 0 {
-                            self.loadedPosts.append(contentsOf: posts)
-                        } else {
-                            self.loadedPosts = posts
-                        }
-                    }
-                    return
-                case .failure(let error):
-                    Amplify.log.error("\(#function) Error loading posts - \(error.localizedDescription)")
-                    DispatchQueue.main.async {
-                        self.photoSharingError = error
-                        self.hasError = true
+            do {
+                let posts = try await dataStoreService.query(Post.self,
+                                                             where: predicateInput,
+                                                             sort: sortInput,
+                                                             paginate: paginationInput)
+                await MainActor.run {
```

Comment:
why does this code block need to be executed in the main dispatch queue but not `self.photoSharingError` and `self.hasError`?